codec-http2/src/main/java/io/netty/handler/codec/http2/ClearTextHttp2ServerUpgradeHandler.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http2;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerAdapter;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.http.HttpServerCodec;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler.UpgradeCodecFactory;
+
+import java.util.List;
+
+import static io.netty.buffer.Unpooled.unreleasableBuffer;
+import static io.netty.handler.codec.http2.Http2CodecUtil.connectionPrefaceBuf;
+
+/**
+ * Providing how starting http2 protocol, by h2c HTTP upgrade, or by h2 with Prior Knowledge.
+ * This handler will set up {@link PriorKnowlegeHandler}, {@link HttpServerCodec},
+ * and {@link UpgradeCodecFactory} into the pipeline when handler added.
+ * And will set up pipeline correctly once it detect the protocol is h2c or h2.
+ */
+public class ClearTextHttp2ServerUpgradeHandler extends ChannelHandlerAdapter {
+    private static final ByteBuf CONNECTION_PREFACE = unreleasableBuffer(connectionPrefaceBuf());
+
+    private final HttpServerCodec httpServerCodec;
+    private final ChannelHandler priorKnowledgeUpgradeHandler;
+    private final HttpServerUpgradeHandler httpServerUpgradeHandler;
+
+    public ClearTextHttp2ServerUpgradeHandler(HttpServerCodec httpServerCodec,
+                                              HttpServerUpgradeHandler httpServerUpgradeHandler,
+                                              ChannelHandler priorKnowledgeUpgradeHandler) {
+        this.httpServerCodec = httpServerCodec;
+        this.httpServerUpgradeHandler = httpServerUpgradeHandler;
+        this.priorKnowledgeUpgradeHandler = priorKnowledgeUpgradeHandler;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.pipeline()
+           .addBefore(ctx.name(), null, new PriorKnowlegeHandler())
```

Comment:
here and elsewhere, this seems to rely on the assumption that ctx.name() isn't named null.  can we do it without using addBefore?